aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1720,21 +1732,37 @@ static Tensor& _linalg_norm_vector_out(Tensor& result, const Tensor& self, optio
     TORCH_CHECK(unique_dims, "Expected dims to be different, got this instead: (", dim, ")");
   }
   if (!case_was_overridden) {
-    if (opt_dtype.has_value()) {
-      result_ = at::norm(self.to(opt_dtype.value()), opt_ord, dim, keepdim);
+    if (flatten) {
+      if (opt_dtype.has_value()) {
+        result_ = at::norm(self.to(opt_dtype.value()), opt_ord.value_or(2));
+      } else {
+        result_ = at::norm(self, opt_ord.value_or(2));
+      }
+      if (keepdim) {
+        result_ = result_.reshape(std::vector<int64_t>(self.dim(), 1));
+      }
     } else {
-      result_ = at::norm(self, opt_ord, dim, keepdim);
+      if (opt_dtype.has_value()) {
+        result_ = at::norm(self.to(opt_dtype.value()), opt_ord, dim, keepdim);
+      } else {
+        result_ = at::norm(self, opt_ord, dim, keepdim);
+      }
     }
   }
   resize_output(result, result_.sizes());
   result.copy_(result_);
   return result;
 }
 
-static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional<Scalar> opt_num_ord, optional<std::string> opt_str_ord, optional<IntArrayRef> opt_dim, bool keepdim, optional<ScalarType> opt_dtype) {
+static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional<Scalar> opt_num_ord, optional<std::string> opt_str_ord, optional<IntArrayRef> opt_dim, bool keepdim, optional<ScalarType> opt_dtype, bool flatten) {
```

Comment:
So this is public API with incompatible options (keepdim, dim on one hand, and flatten on another). That's not great.

Reply:
`keepdim` is compatible with the `flatten` argument, it will just restore all the dimensions that were flattened, which is also what happens with `torch.norm`. In other words, `torch.norm(a, keepdim=True)` will give the same result as `torch.linalg.norm(a, keepdim=True, flatten=True)`. However, perhaps the word "flatten" is not ideal, since it implies a direct parallel with `torch.flatten`--`torch.flatten` and `keepdim` don't seem like compatible concepts. Maybe this issue can be resolved by renaming the `flatten` argument to `alldims`, `allreduce`, or something like that.

You're right that `dim` and `flatten` are incompatible, and this PR currently has a check to ensure that if `flatten == True`, then `dim` must be None. I thought that this check would be sufficient to resolve this incompatibility, but I understand that it is a bit of a convoluted API.

Maybe there is a simpler solution to solve the main problem of this PR: how do we offer an N-D vector norm in `torch.linalg.norm`, while still being numpy compatible? If there is a better solution, I'm happy to implement it. What are your thoughts?

Perhaps one idea could be to offer a new public function `torch.linalg.vector_norm` instead of the `flatten` argument.

Reply:
Yeah, we've been discussing this offline with @mruberry, and having a separate function for vector_norm is a great idea. It will preserve clean UX of torch.linalg.norm and provide existing functionality of torch.norm. 
We think that `torch.linalg.vector_norm` should mimic `torch.norm` behavior, except, obviously, it should not support matrix norms at all. Can you please open an issue describing this proposal? Thanks!

Reply:
Thank you both, @kurtamohler and @ngimel, for the UX thinking, and your flexibility here, @kurtamohler. I think torch.linalg.vector_norm is likely to be more discoverable and offer clearer behavior to users. Looking forward to following-up on the issue.